hawkmoth/parser.py

```diff
@@ -239,6 +239,15 @@ def _recursive_parse(comments, cursor, nest, compat):
 
     return [(doc, meta)]
 
+def _clang_diagnostics(tu):
+    errors = []
+    for d in tu.diagnostics:
+        # Filter for warnings and errors
+        if d.severity >= 2:
```

Comment:
I don't know if we should be passing so many things back to the user. This will give warnings like 'symbol not defined'. Well... we shouldn't replace the build system. If we call the directive with a single source file, it's only natural a bunch of definitions will be missing.

Haven't found documentation on what each level means though.

Reply:
Yep. You're right. Documentation is 0 :)

Here's what sources says:
```
    Ignored = 0
    Note    = 1
    Warning = 2
    Error   = 3
    Fatal   = 4
```

Reply:
I disagree. I'd go for error/fatal only, whichever relates to Clang bailing out on us altogether. E.g. 'missing definitions' is to be expected under normal operation which renders the output uselessly noisy at level 2.

Sphinx should be printing errors found in documentation (function parameters missing in the docstring or not matching, or even more fundamental bad uses). I think it's good to print something about Clang failing completely, but the user should have actually built the code to find source errors.

I just don't know what's the difference between error and fatal exactly.

Reply:
Fatal is missing files. When doing the AST it recursively reads header files. If it can't find it it will go with severity 4 (Fatal).

Reply:
BTW, when this happens the AST parser will silently stop processing symbols. The end result is that we end up comments in the documentation but no cursor (symbol declaration).

Reply:
Hmm, maybe I'm wrong then. Plenty of our test cases are incomplete sources and yet they produce good documents though. I assumed missing headers would have a similar result, but haven't actually tried it.

For instance, I wouldn't want to do this in the documentation if possible:
```rst
.. c:autodoc:: source.c
   :clang: -Ipath/to/headers
```

Reply:
Yep, that would be need for some of the tests. See my initial PR where I've modified a bit the test scripts and how it prints.

Reply:
I've tested a bit the sphinx methods of logging/printing:

- info prints without any prefix 
- debug doesn't print all all (I assume that this has to be enabled)
- critical/fatal/warning methods all add a prefix/prepend with `WARNING: ` string before. I'd expected these to print different but it doesn't appear to be the case?

Reply:
Crap... If there's no way around it then I think we must add all those warnings and errors of course. The user should know why the documentation build failed.

Maybe there's something in how we invoke Clang that can make it ignore missing files. If we cut the parsing at different levels it may affect how much information we can extract from the code though, so I'm not very optimistic.

In the meantime I'm closing this PR, I think it served its purpose anyway and you've kept updating your own. Let's consolidate the discussion over there.